### PR TITLE
Scripts for integration with DL4RT's nemo-editor

### PR DESCRIPTION
**Pull request by tkornuta-nvidia**
- [x] get_collections.py - generates a JSON file with list of collections

Exemplary content of `collections.json`:
```
[
{"id": "asr", "name": "nemo.collections.asr", "description": "Automatic Speech Recognition collection", "version": "0.88.1b0", "author": "NVIDIA Corporation"}, 
{"id": "common", "name": "nemo.collections.common", "description": "Common collection", "version": "0.88.1b0", "author": "NVIDIA Corporation"}, 
{"id": "cv", "name": "nemo.collections.cv", "description": "Computer Vision collection", "version": "0.88.1b0", "author": "NVIDIA Corporation"}, 
{"id": "nlp", "name": "nemo.collections.nlp", "description": "Natural Language Processing collection", "version": "0.88.1b0", "author": "NVIDIA Corporation"}, 
{"id": "tts", "name": "nemo.collections.tts", "description": "Text-To-Speech collection", "version": "0.88.1b0", "author": "NVIDIA Corporation"}]
```

 - [x] get_modules.py - generates a JSON file with list of modules from a given collection (the script mocks up module's arguments, input_types and output_types)

Exemplary content of `modules.json` for `python get_modules.py --c cv`:
```
[
{"name": "MNISTDataset", "cls": "<class 'nemo.collections.cv.datasets.mnist_dataset.MNISTDataset'>",
 "arguments": ["jasper", "activation", "feat_in", "normalization_mode", "residual_mode", "norm_groups", "conv_mask", "frame_splicing", "init_mode"],
 "input_types": {"audio_signal": "axes: (batch, dimension, time); elements_type: MelSpectrogramType", "length": "axes: (batch,); elements_type: LengthType"},
 "output_types": {"encoder_output": "axes: (batch, dimension, time); elements_type: AcousticEncodedRepresentation"}}, 
{"name": "LeNet5", "cls": "<class 'nemo.collections.cv.modules.lenet5.LeNet5'>",
 "arguments": ["jasper", "activation", "feat_in", "normalization_mode", "residual_mode", "norm_groups", "conv_mask", "frame_splicing", "init_mode"],
 "input_types": {"audio_signal": "axes: (batch, dimension, time); elements_type: MelSpectrogramType", "length": "axes: (batch,); elements_type: LengthType"},
 "output_types": {"encoder_output": "axes: (batch, dimension, time); elements_type: AcousticEncodedRepresentation"}}, 
{"name": "NLLLoss", "cls": "<class 'nemo.collections.cv.losses.nll_loss.NLLLoss'>",
 "arguments": ["jasper", "activation", "feat_in", "normalization_mode", "residual_mode", "norm_groups", "conv_mask", "frame_splicing", "init_mode"],
 "input_types": {"audio_signal": "axes: (batch, dimension, time); elements_type: MelSpectrogramType", "length": "axes: (batch,); elements_type: LengthType"},
 "output_types": {"encoder_output": "axes: (batch, dimension, time); elements_type: AcousticEncodedRepresentation"}}]
```

### CODE REVIEW

File: external/get_collections.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+
+
+def analyse_collection(id, col):
+    return {
+        "id": id,
+        "name": col.__name__,
+        "description": col.__description__,
+        "version": col.__version__,
+        "author": col.__author__
```
Comment (titu1994) [inline external/get_collections.py:41]: Why do we need author? Almost entire team works on multiple parts of collections so author is not particularly appropriate here.

File: external/get_collections.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    colletions_dir = os.path.dirname(nemo.collections.__file__)
+    logging.info('Analysing collections in `{}`'.format(colletions_dir))
+
+    # Get list dir.
+    #for sub_dir in os.listdir(colletions_dir):
+    #    print(sub_dir)
+
+    col_list = []
+    # Iterate over all collections.
+    for id, col in zip(["asr", "common", "cv", "nlp", "tts"],[col_asr, col_common, col_cv, col_nlp, col_tts]):
```
Comment (titu1994) [inline external/get_collections.py:62]: Maybe refactor the string list of names into a dictionary pointing to import paths ? And then use import lib to import those ?

File: nemo/collections/__init__.py [hunk 3]
```diff
@@ -11,3 +11,5 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+
+# __all__ = ["asr", "common", "cv", "nlp", "tts"]
```
Comment (titu1994) [inline nemo/collections/__init__.py:15]: Revert. This if uncommented requires all Nemo collections to be installed all the time.

File: nemo/collections/asr/__init__.py [hunk 4]
```diff
@@ -13,3 +13,14 @@
...
 
 from nemo.collections.asr import data, losses, models, modules
+
+from nemo.package_info import __version__
+
+# Set collection version equal to NeMo version.
+__version = __version__
+
+# Authorship.
+__author__ = "NVIDIA Corporation"
```
Comment (titu1994) [inline nemo/collections/asr/__init__.py:23]: Authorship makes very little sense here, i suggest simply removing the field entirely

File: external/get_collections.py (hunk 1 above)
Comment (tkornuta-nvidia) [reply external/get_collections.py:41]: I agree, that is why I added __authors__ = "NVIDIA Corporation" to every collection's `__init__.py`.

Probably will remove this, for now I am just "fulfilling" the requirements that Chris send:

"Script 1: If NeMo has collection metadata (human-readable descriptions, URLs, version strings…) we should retrieve that as well.
Proposed format is an array of dictionaries: [ { “id” : “Collection1” }, { “id” : “Collection2” }, … ]
If more information is available, add additional fields to the dictionary: { “id” : “Collection1”, **“author” : “NVIDIA Corporation”**, … }
We can also move metadata to script 2 if that’s more convenient."

File: nemo/collections/__init__.py (hunk 3 above)
Comment (tkornuta-nvidia) [reply nemo/collections/__init__.py:15]: removed

File: nemo/collections/asr/__init__.py (hunk 4 above)
Comment (tkornuta-nvidia) [reply nemo/collections/asr/__init__.py:23]: discussed this above, let's see what DL4RT will say.

File: external/get_collections.py (hunk 2 above)
Comment (tkornuta-nvidia) [reply external/get_collections.py]: Thanks, @titu1994, importlib worked like a charm! 👍

File: external/get_collections.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
+# -*- coding: utf-8 -*-
```
Comment (okuchaiev) [inline external/get_collections.py:1]: please paste correct licence header from dev guide

File: external/get_modules.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
```
Comment (okuchaiev) [inline external/get_modules.py:1]: same for licence

File: nemo/collections/asr/__init__.py (hunk 4 above)
Comment (okuchaiev) [reply nemo/collections/asr/__init__.py:22]: I think I missed that discussion, can you please point us to it?

Comment (tkornuta-nvidia) [reply nemo/collections/asr/__init__.py:22]: Chris explicitly wrote in email from Friday, June 26, 2020 4:40 PM:
"Script 1: If NeMo has collection metadata (human-readable descriptions, URLs, version strings…) we should retrieve that as well.
Proposed format is an array of dictionaries: [ { “id” : “Collection1” }, { “id” : “Collection2” }, … ]
If more information is available, add additional fields to the dictionary: { “id” : “Collection1”, “author” : “NVIDIA Corporation”, … }"

The discussion "what kind of metadata per collection" we want to send to nemo_editor is open, added _some fields_ to collection's `__init__.py` really as a proof of concept...

File: external/get_collections.py (hunk 5 above)
Comment (tkornuta-nvidia) [reply external/get_collections.py:1]: Ok, updated "NVIDIA" to "NVIDIA CORPORATION".

File: external/get_modules.py (hunk 6 above)
Comment (tkornuta-nvidia) [reply external/get_modules.py:1]: Ok, updated "NVIDIA" to "NVIDIA CORPORATION".

File: external/get_collections.py (hunk 5 above)
Comment (okuchaiev) [reply external/get_collections.py:1]: Please just copy-paste the license.
We don't use  # ======= anywhere in candidate. Should be really just the same license header everywhere

Comment (tkornuta-nvidia) [reply external/get_collections.py:1]: ? 

This is just a decorator. It doesn't have anything to do with the content of the header - which is 100% the same, точка в точку.

I honestly do not think we should enforce any user in any way he/she wants to _comment_ his/her code.  Hope we agree on that.

Aside of that, in our dev guide we also do not have `# -*- coding: utf-8 -*-` - so does it means that it _bad_?

**Comment by tkornuta-nvidia** [comment]
> Why have "NeMo/external"? why don't put those two scripts under "NeMo/scripts"?

I didn't know where to put them. 

But not sure about `scripts`. What other scripts will be there? For now I see some asr preprocessing. Are you sure you want to put all the scripts into the same _bucket_?

How about `scripts/external/` or e.g. `scripts/nemo_editor/`?

**Pull request merged**